### ISSUE TITLE
Protected branches

### ISSUE THREAD

**Issue by VasifAghayev** [egnyte/gitlabform#510]
Hi,

I have the following configuration related to protected branches in config.yaml file: 
```
main:
     protected: true
     allowed_to_merge:
        - access_level: developer
     allowed_to_unprotect:
        - access_level: maintainer
     allowed_to_push:
        - user: user_name
```
However, when I checked protected branches via API, I got the following response:
```
 {
    "id": 186414,
    "name": "main",
    "push_access_levels": [
      {
        "access_level": 40,
        "access_level_description": "some-description",
        "user_id": user_id,
        "group_id": null
      }
    ],
    ...
```
Why is the value of "access_level" parameter not null? Is it expected behavior? Maybe I missing something because I thought with the above-mentioned configuration the value of access_level should be null

**Comment by gdubicki** [comment]
Hey @VasifAghayev! Thanks for reporting. That does look like a bug on a first glance. I’ll test it and get back to you soon.

**Comment by VasifAghayev** [comment]
Thanks! Just for FYI, GitLab version is 14.10.5

**Comment by gdubicki** [comment]
I tried to reproduce the problem with a simple test linked above, but as you can see the test was passing.

However after I dug deeper I noticed that a helper function in our tests code tries to get the access levels for push/merge/unprotect only if there is no `user_id` provided: https://github.com/gitlabform/gitlabform/blob/main/tests/acceptance/__init__.py#L184-L187

And when I checked the actual request and response I noticed that the access level for push is set in the response, although it's NOT provided in the request. Note that **this is GitLab's behavior**, NOT GitLabForm:

`http://localhost:80 "POST /api/v4/projects/gitlabform_group_likeness%2Fgitlabform_project_anagram/protected_branches?name=gitlabform_branch_amperage HTTP/1.1" 201 514`

`===> json = {"allowed_to_merge": [{"access_level": 30}], "allowed_to_push": [{"user_id": 28}], "allowed_to_unprotect": [{"access_level": 40}]}`

`<--- json = {"allow_force_push": false, "code_owner_approval_required": false, "id": 32, "merge_access_levels": [{"access_level": 30, "access_level_description": "Developers + Maintainers", "group_id": null, "id": 36, "user_id": null}], "name": "gitlabform_branch_amperage", "push_access_levels": [{"access_level": 40, "access_level_description": "gitlabform_user_luxury_1 Example", "group_id": null, "id": 42, "user_id": 28}], "unprotect_access_levels": [{"access_level": 40, "access_level_description": "Maintainers", "group_id": null, "id": 12, "user_id": null}]}`

I think that this default, automatically set access level is to indicate that Maintainers can push to protected branches as well:

<img width="809" alt="Screenshot 2023-04-02 at 20 17 24" src="https://user-images.githubusercontent.com/566632/229374155-740bf458-c576-4920-a44c-635eebd680bd.png">

(screenshot from the [GitLab docs](https://docs.gitlab.com/ee/user/permissions.html#project-members-permissions))

This explanation matches [the example provided in GitLab's Protected Branches API docs](https://docs.gitlab.com/ee/api/protected_branches.html#example-with-user--group-level-access), where allowing to push only for an admin user sets the access level to `null`, but the access level description to a string `Administrator`.

Also in the GitLab's Terraform provider docs the access level and access level description fields are documented as "read only" under the "allowed to push" if you want to set it to a user id (or a group id):

<img width="620" alt="Screenshot 2023-04-02 at 20 24 37" src="https://user-images.githubusercontent.com/566632/229374597-3924172e-2a5a-4f01-9dba-de80ab672c18.png">

(screenshot from the [provider's docs](https://registry.terraform.io/providers/gitlabhq/gitlab/latest/docs/resources/branch_protection#nestedblock--allowed_to_push))

To sum it up: this is just how this API works in GitLab...

What I am thinking about now is if we can document how to use this section better, to make this less surprising.

**Comment by VasifAghayev** [comment]
Hi gdubicki
Sorry for the late reply. Thanks a lot for the explanation and testing. I have just one additional question: 
Do I understand correctly, that we can't explicitly set `access_level` to `null`, when `user_id` is provided (in the config file/API request) for protected branches, due to how the GitLab API is currently implemented?
